File: apps/clojure/src/fizzbuzz/application/command.clj
(ns fizzbuzz.application.command
  (:require [fizzbuzz.domain.model :as model]
            [fizzbuzz.domain.type :as type]))

;; ── プロトコル: FizzBuzzCommand ──────────────────────────

(defprotocol FizzBuzzCommand
  "FizzBuzz コマンドプロトコル"
  (execute [this]
    "コマンドを実行する"))

;; ── 単一値コマンド ───────────────────────────────────────

(defrecord FizzBuzzValueCommand [number type-num]
  FizzBuzzCommand
  (execute [_this]
    (let [fizz-buzz-type (type/create-type type-num)
          value (model/create-fizz-buzz-value number)]
      (type/generate-string fizz-buzz-type value))))

;; ── リストコマンド ───────────────────────────────────────

(defrecord FizzBuzzListCommand [numbers type-num]
  FizzBuzzCommand
  (execute [_this]
    (let [fizz-buzz-type (type/create-type type-num)]
      (->> numbers
           (mapv model/create-fizz-buzz-value)
           (mapv #(type/generate-string fizz-buzz-type %))
           model/create-fizz-buzz-list))))

File: apps/scala/src/main/scala/fizzbuzz/application/Command.scala
package fizzbuzz.application

import fizzbuzz.domain.{FizzBuzzList, FizzBuzzType, FizzBuzzValue}

trait FizzBuzzCommand:
  def execute(): String

case class FizzBuzzValueCommand(number: Int, typeNumber: Int) extends FizzBuzzCommand:
  def execute(): String =
    val fizzBuzzType = FizzBuzzType.create(typeNumber)
    val value = FizzBuzzValue(number, fizzBuzzType.generate(number))
    value.toString

case class FizzBuzzListCommand(count: Int, typeNumber: Int) extends FizzBuzzCommand:
  def execute(): String =
    val fizzBuzzType = FizzBuzzType.create(typeNumber)
    FizzBuzzList.create(count, fizzBuzzType).toStringList.mkString("\n")
